File: clients/java/src/test/java/com/azrtydxb/novamem/ScenarioTest.java
package com.azrtydxb.novamem;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.time.Duration;
import java.util.Locale;
import java.util.concurrent.CancellationException;
import java.util.concurrent.CompletionException;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

/** The shared behaviour suite (clients/contract/scenarios.json, ADR 0009). */
class ScenarioTest {
  static final Path CONTRACT = Path.of("..", "contract").toAbsolutePath().normalize();
  static JsonNode scen;
  static Process proc;
  static String url;
  static String closed;

  @BeforeAll
  static void start() throws Exception {
    scen = Json.MAPPER.readTree(CONTRACT.resolve("scenarios.json").toFile());
    proc =
        new ProcessBuilder("sh", "scenario-server.sh", "-scenarios", "scenarios.json")
            .directory(CONTRACT.toFile())
            .redirectError(ProcessBuilder.Redirect.INHERIT)
            .start();
    String[] parts =
        new BufferedReader(new InputStreamReader(proc.getInputStream(), StandardCharsets.UTF_8))
            .readLine()
            .split(" ");
    url = parts[1];
    closed = parts[2].split("=")[1];
  }

  @AfterAll
  static void stop() {
    proc.destroyForcibly();
  }

  static Stream<String> ids() throws Exception {
    JsonNode s = Json.MAPPER.readTree(CONTRACT.resolve("scenarios.json").toFile()).get("scenarios");
    return StreamSupport.stream(s.spliterator(), false).map(x -> x.get("id").asText());
  }

  static boolean isEmpty(Object r) {
    JsonNode n = Json.MAPPER.valueToTree(r);
    return (n.isArray() && n.isEmpty())
        || (n.isObject() && n.has("results") && n.get("results").isEmpty());
  }

  /** "" when every field of want is in got with the same value; else where they differ. */
  static String subset(JsonNode want, JsonNode got, String path) {
    if (want.isObject()) {
      if (got == null || !got.isObject()) {
        return path + ": want an object, got " + got;
      }
      for (var e : want.properties()) {
        String d = subset(e.getValue(), got.get(e.getKey()), path + "." + e.getKey());
        if (!d.isEmpty()) {
          return d;
        }
      }
      return "";
    }
    if (want.isNumber() && got != null && got.isNumber()) {
      return want.doubleValue() == got.doubleValue()
          ? ""
          : path + ": got " + got + ", want " + want;
    }
    return want.equals(got) ? "" : path + ": got " + got + ", want " + want;
  }

  // proved by: removing the degraded-empty check from Client.searchAsync fails
  // search-degraded-empty-is-unavailable; removing the redaction in Transport
  // fails token-echoed-in-401-is-redacted.
  @ParameterizedTest
  @MethodSource("ids")
  void testScenarios(String id) throws Exception {
    JsonNode s = null;
    for (JsonNode x : scen.get("scenarios")) {
      if (x.get("id").asText().equals(id)) {
        s = x;
      }
    }
    JsonNode call = s.get("call");
    String token = scen.get("token").asText();
    Object result = null;
    Throwable err = null;
    if (call.get("class").asText().equals("ctor")) {
      JsonNode a = call.get("args");
      try {
        new Client(
            NovamemConfig.builder()
                .baseUrl(a.get("baseUrl").asText().replace("<server>", url))
                .token(a.get("token").asText())
                .build());
      } catch (Exception e) {
        err = e;
      }
    } else {
      String base =
          s.path("respond").toString().contains("\"refused\"")
              ? "http://127.0.0.1:" + closed + "/s/" + id
              : url + "/s/" + id;
      NovamemConfig cfg =
          NovamemConfig.builder()
              .baseUrl(base)
              .token(token)
              .timeout(Duration.ofMillis(scen.get("timeoutMs").asLong()))
              .build();
      Dispatch.Clients c =
          new Dispatch.Clients(new Client(cfg), new Management(cfg), new Admin(cfg));
      boolean cancel = call.has("cancelAfterMs");
      try {
        result =
            Dispatch.TABLE
                .get(call.get("method").asText())
                .call(c, call.get("args"), cancel, cancel ? call.get("cancelAfterMs").asLong() : 0);
      } catch (CompletionException e) {
        err = e.getCause();
      } catch (Exception e) {
        err = e;
      }
    }
    String outcome =
        err == null
            ? (isEmpty(result) ? "empty" : "ok")
            : err instanceof CancellationException
                ? "canceled"
                : err instanceof NovamemException ne && ne.isUnavailable()
                    ? "unavailable"
                    : err instanceof NovamemException ne2 && ne2.isNotFound()
                        ? "not_found"
                        : "error";
    JsonNode exp = s.get("expect");
    assertEquals(exp.get("outcome").asText(), outcome, id + ": " + err);
    if (err != null) {
      assertFalse(String.valueOf(err).contains(token), id + ": token leaked");
      if (err instanceof NovamemException ne) {
        if (exp.has("retryable")) {
          assertEquals(exp.get("retryable").asBoolean(), ne.isRetryable(), id + ": retryable");
        }
        if (exp.has("statusCode")) {
          assertEquals(exp.get("statusCode").asInt(), ne.statusCode(), id + ": status");
        }
        if (exp.has("code")) {
          assertEquals(exp.get("code").asText(), ne.code(), id + ": code");
        }
      }
      if (exp.has("messageContains")) {
        String want = exp.get("messageContains").asText().toLowerCase(Locale.ROOT);
        assertTrue(
            String.valueOf(err.getMessage()).toLowerCase(Locale.ROOT).contains(want),
            id + ": message " + err.getMessage());
      }
    }
    if (exp.has("result")) {
      String d = subset(exp.get("result"), Json.MAPPER.valueToTree(result), "$");
      assertEquals("", d, id + ": result");
    }
    if (!call.get("class").asText().equals("ctor")) {
      HttpResponse<String> v =
          HttpClient.newHttpClient()
              .send(
                  HttpRequest.newBuilder(URI.create(url + "/_verdict/" + id)).build(),
                  HttpResponse.BodyHandlers.ofString());
      JsonNode verdict = Json.MAPPER.readTree(v.body());
      assertEquals(0, verdict.get("mismatches").size(), id + ": " + verdict);
      if (s.path("expectRequest").isArray() && s.get("expectRequest").isEmpty()) {
        assertEquals(0, verdict.get("requests").asInt(), id + ": expected no request");
      }
    }
  }
}

File: clients/php/tests/ScenarioTest.php
<?php

// The shared behaviour suite (clients/contract/scenarios.json, ADR 0009).

declare(strict_types=1);

use Novamem\Admin;
use Novamem\Client;
use Novamem\Config;
use Novamem\Management;
use Novamem\NovamemException;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;

final class ScenarioTest extends TestCase
{
    /** @var resource|null */
    private static $proc = null;
    private static string $url = "";
    private static string $closed = "";
    /** @var array<string, mixed> */
    private static array $scen = [];

    private static function contract(): string
    {
        return dirname(__DIR__, 2) . "/contract";
    }

    public static function setUpBeforeClass(): void
    {
        self::$scen = json_decode(
            (string) file_get_contents(self::contract() . "/scenarios.json"),
            true,
        );
        self::$proc = proc_open(
            ["sh", "scenario-server.sh", "-scenarios", "scenarios.json"],
            [1 => ["pipe", "w"]],
            $pipes,
            self::contract(),
        );
        [, self::$url, $closed] = explode(" ", trim((string) fgets($pipes[1])));
        self::$closed = explode("=", $closed)[1];
    }

    public static function tearDownAfterClass(): void
    {
        if (self::$proc !== null) {
            proc_terminate(self::$proc, 9);
        }
    }

    /** @return array<string, array{string}> */
    public static function ids(): array
    {
        $s = json_decode(
            (string) file_get_contents(self::contract() . "/scenarios.json"),
            true,
        );
        $out = [];
        foreach ($s["scenarios"] as $x) {
            $out[$x["id"]] = [$x["id"]];
        }
        return $out;
    }

    private static function wire(mixed $r): mixed
    {
        return is_object($r) && method_exists($r, "toArray")
            ? $r->toArray()
            : $r;
    }

    private static function isEmptyResult(mixed $r): bool
    {
        $w = self::wire($r);
        return $w === [] ||
            (is_array($w) &&
                array_key_exists("results", $w) &&
                $w["results"] === []);
    }

    private static function subset(
        mixed $want,
        mixed $got,
        string $path = '$',
    ): string {
        if (is_array($want) && !array_is_list($want)) {
            if (!is_array($got)) {
                return "$path: want an object, got " . json_encode($got);
            }
            foreach ($want as $k => $v) {
                $d = self::subset($v, $got[$k] ?? null, "$path.$k");
                if ($d !== "") {
                    return $d;
                }
            }
            return "";
        }
        return $want === $got
            ? ""
            : "$path: got " .
                    json_encode($got) .
                    ", want " .
                    json_encode($want);
    }

    // proved by: removing the degraded-empty check in Client::search fails
    // search-degraded-empty-is-unavailable; removing the redaction fails
    // token-echoed-in-401-is-redacted.
    #[DataProvider("ids")]
    public function testScenarios(string $id): void
    {
        $s = null;
        foreach (self::$scen["scenarios"] as $x) {
            if ($x["id"] === $id) {
                $s = $x;
            }
        }
        if (in_array("cancel", $s["requires"] ?? [], true)) {
            $this->markTestSkipped(
                "the PHP client has no cancellation primitive: $id",
            );
        }
        $call = $s["call"];
        $token = self::$scen["token"];
        $r = null;
        $e = null;
        if ($call["class"] === "ctor") {
            try {
                new Client(
                    new Config(
                        str_replace(
                            "<server>",
                            self::$url,
                            $call["args"]["baseUrl"],
                        ),
                        $call["args"]["token"],
                    ),
                );
            } catch (\InvalidArgumentException $err) {
                $e = $err;
            }
        } else {
            $base = str_contains(
                (string) json_encode($s["respond"] ?? null),
                '"refused"',
            )
                ? "http://127.0.0.1:" . self::$closed . "/s/$id"
                : self::$url . "/s/$id";
            $cfg = new Config($base, $token, self::$scen["timeoutMs"]);
            $clients = [
                "Client" => new Client($cfg),
                "Management" => new Management($cfg),
                "Admin" => new Admin($cfg),
            ];
            try {
                $r = Dispatch::table()[$call["method"]](
                    $clients,
                    $call["args"],
                );
            } catch (NovamemException $err) {
                $e = $err;
            }
        }
        $outcome = match (true) {
            $e === null => self::isEmptyResult($r) ? "empty" : "ok",
            $e instanceof NovamemException && $e->isUnavailable()
                => "unavailable",
            $e instanceof NovamemException && $e->isNotFound() => "not_found",
            default => "error",
        };
        $exp = $s["expect"];
        $this->assertSame(
            $exp["outcome"],
            $outcome,
            "$id: " . ($e?->getMessage() ?? ""),
        );
        if ($e !== null) {
            $this->assertStringNotContainsString($token, (string) $e);
            // The exception's own fields, as the other SDKs check repr/inspect.
            // (print_r($e) would dump the whole trace, which holds PHPUnit's
            // test case and, with it, this scenario's token.)
            if ($e instanceof NovamemException) {
                $fields = [
                    $e->op(),
                    $e->statusCode(),
                    $e->errorCode(),
                    $e->detail(),
                    $e->getMessage(),
                ];
                $this->assertStringNotContainsString(
                    $token,
                    print_r($fields, true),
                );
            }
            if ($e instanceof NovamemException) {
                if (array_key_exists("retryable", $exp)) {
                    $this->assertSame($exp["retryable"], $e->isRetryable());
                }
                if (array_key_exists("statusCode", $exp)) {
                    $this->assertSame($exp["statusCode"], $e->statusCode());
                }
                if (array_key_exists("code", $exp)) {
                    $this->assertSame($exp["code"], $e->errorCode());
                }
            }
            if (array_key_exists("messageContains", $exp)) {
                $this->assertStringContainsStringIgnoringCase(
                    $exp["messageContains"],
                    $e->getMessage(),
                );
            }
        }
        if (array_key_exists("result", $exp)) {
            $this->assertSame("", self::subset($exp["result"], self::wire($r)));
        }
        if ($call["class"] !== "ctor") {
            $v = json_decode(
                (string) file_get_contents(self::$url . "/_verdict/$id"),
                true,
            );
            $this->assertSame([], $v["mismatches"]);
            if (($s["expectRequest"] ?? null) === []) {
                $this->assertSame(0, $v["requests"]);
            }
        }
    }
}
